New documentation, installation and setup scripts (work in progress)

meyerj opened the pull request:
This PR adds a README.md, a new installation manual INSTALL.md, a installation script and a collection of environment setup scripts to the root folder of the orocos_toolchain repository.

The installation instructions have been partially copied from https://github.com/orocos-toolchain/rtt/blob/master/doc/xml/orocos-installation.xml. The new script `install.sh` does not depend on catkin or Ruby and is able to clone the orocos_toolchain, log4cpp, rtt and ocl repositories and build and install these packages in order. Automatic installation of dependencies is out of scope and should be left to the user, based on good documentation.

The `env.sh` script has been replaced by a wrapper script which executes its arguments with a proper environment. The environment itself is managed by the new `setup.sh` script, which can be sourced from any shell and
- sets the `OROCOS_INSTALL_PREFIX` and `OROCOS_TARGET` environment variables
- prepends the respective Orocos installation directories to `PATH`, `RTT_COMPONENT_PATH` and `PKG_CONFIG_PATH` while avoiding duplicates if the script is sources more than once
- sources additional package-specific hooks in `${OROCOS_INSTALL_PREFIX}/etc/orocos/${OROCOS_TARGET}/profile.d/*.sh`, e.g. to setup the `LUA_PATH` variables in OCL

See also related PRs
- https://github.com/orocos-toolchain/rtt/pull/160 in RTT and
- https://github.com/orocos-toolchain/ocl/pull/43 in OCL.

We do not include orogen and its dependencies for now and would refer to the [Rock](http://www.rock-robotics.org/stable/index.html) installation instructions instead. If someone else wants to add orogen support to the script, feel free to do so, but it should always be an option to install these additional packages.

We want to stay compatible with whatever build and installation tools already exist. So please comment if the renaming of the `env.sh` script to `setup.sh` would be a problem for you. Other than that, it should stay compatible if Ruby-specific settings are moved to env-hooks in the respective packages.

meyerj committed:
Added initial version of installation script and manual (work in progress)

meyerj committed:
Fully working version of the installation script, added env-hooks and README.md

meyerj committed:
Allow to source the env.sh script to stay backwards-compatible

meyerj committed:
install.sh: Fixed --url command line option and build in current working directory

meyerj committed:
setup.sh: also update and export the CMAKE_PREFIX_PATH

meyerj committed:
orocos_toolchain: find OROCOS-RTT before configuring the setup scripts

Commit:
Add travis script to test installation script

Commit:
Silly update to trigger travis

Commit:
Remove comments

Commit:
Fix travis file extension

Commit:
Fix apt dependency

smits commented on install.sh:
```diff
@@ -0,0 +1,477 @@
+      -DCMAKE_INSTALL_PREFIX="${PREFIX}" \
+      "${FEATURE_CMAKE_ARGS[@]}" "${CMAKE_ARGS[@]}"
+    do_run make ${JOBS} "${MAKE_ARGS[@]}"
+    do_run make install
```
Since the install prefix is by default /usr/local, and normal users do not have write rights by default to that folder, it might be ok to add sudo here?

meyerj replied:
I would not add the sudo prefix in the script itself. If users do not overwrite the install location with the `--prefix` option, the whole install.sh script should run as super-user. We could update the INSTALL.md file accordingly.

Commit:
Change prefix to a writable place and enable all features

Commit:
Remove compiler matrix as it doesn't work properly

meyerj committed:
Merge branch 'toolchain-2.9' into smits-patch-1

Commit:
Only use containerized builds

meyerj committed:
Fix readlink -f on MacOS X

meyerj committed:
install.sh: do not call lsb_release on non-Linux platforms

smits commented on INSTALL.md:
```diff
@@ -0,0 +1,166 @@
+mkdir -p ~/orocos-toolchain
+cd ~/orocos-toolchain
+wget https://raw.githubusercontent.com/orocos-toolchain/orocos_toolchain/master/install.sh
+./install.sh --help
```
The downloaded script is not executable by default, so either we prepend this with ```sh``` or we add a ```chmod +x ./install.sh``` step.

meyerj replied:
Either that or prefix the script with `bash`:
```
bash install.sh --help
```

Or shortcut that downloads and runs the script in a single command is:
```
wget -qO - https://raw.githubusercontent.com/orocos-toolchain/orocos_toolchain/master/install.sh | bash -s -- [options]
```
(motivated by https://get.docker.com/, we could create a similar alias as http://get.orocos.org/)

Commit:
Merge pull request #16 from orocos-toolchain/smits-patch-1

Preliminary merge

Commit:
Add missing dependency if reinstall if required

Signed-off-by: Ruben Smits <ruben@intermodalics.eu>

Commit:
Make travis build verbose

Signed-off-by: Ruben Smits <ruben@intermodalics.eu>

Commit:
replace . with /Users/rubensmits/orocos-toolchain for OSX

Signed-off-by: Ruben Smits <ruben@intermodalics.eu>

Commit:
Update RTT

meyerj committed:
install.sh: fixed type and verbose display of features

Commit:
Link gnu-getopt

Signed-off-by: Ruben Smits <ruben@intermodalics.eu>

Commit:
Force linking of gnu-getopt

Signed-off-by: Ruben Smits <ruben@intermodalics.eu>

Commit:
Add apt addons to matrix entry

Signed-off-by: Ruben Smits <ruben@intermodalics.eu>

meyerj committed:
rtt: updated submodule to include Boost 1.62 patches

meyerj committed:
rtt: updated submodule

Commit:
Remove broken xenial build matrix entry

Signed-off-by: Ruben Smits <ruben@intermodalics.eu>

Commit:
Update rtt

Signed-off-by: Ruben Smits <ruben@intermodalics.eu>

meyerj committed:
rtt: updated submodule

meyerj committed:
Merge branch 'toolchain-2.9' into installation-script

meyerj committed:
Look for env-hooks in etc/orocos/profile.d, without the OROCOS_TARGET part

... and install setup files to etc/orocos unconditionally.

Signed-off-by: Johannes Meyer <johannes@intermodalics.eu>

meyerj committed:
Merge branch 'installation-script' into travis-for-install-script

meyerj committed:
update submodule rtt

meyerj committed:
updated submodules rtt and ocl

Commit:
travis: add readline dependency for OCL

Signed-off-by: Ruben Smits <ruben@intermodalics.eu>

meyerj committed:
travis: disable building of tests

Unit tests will be built and run for individual projects, but not for this meta repository. The purpose of this job is mainly to test the installation script.

meyerj committed:
Merge pull request #15 from orocos-toolchain/travis-for-install-script

Travis for install script

meyerj committed:
Merge branch 'toolchain-2.9' into installation-script

mjbogusz commented:
What's the status of this PR?

meyerj commented on INSTALL.md:
```diff
@@ -0,0 +1,166 @@
+| Program / Library               | Minimum Version               | Description                                                                                                                                                                                                                                                                            |
+|---------------------------------|-------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| CMake                           | 2.8.3 (all platforms)         | See resources on cmake.org for pre-compiled packages in case your distribution does not support this version                                                                                                                                                                           |
+| Boost C++ Library               | 1.33.0 (1.40.0 recommended!)  | Boost.org from version 1.33.0 on has a very efficient (time/space) lock-free smart pointer implementation which is used by Orocos. 1.36.0 has boost::intrusive which we require on Windows with MSVS. 1.40.0 has a shared_ptr implementation we require when building Service objects. |
```
Update minimum Boost version to 1.40, check https://github.com/orocos-toolchain/rtt/pull/199.

meyerj commented on INSTALL.md:
```diff
@@ -0,0 +1,166 @@
+| Boost C++ Test Library          | 1.33.0 (During build only)    | Boost.org test library ('unit_test_framework') is required if you build the RTT from source and ENABLE_TESTS=ON (default). The RTT libraries don't depend on this library, it is only used for building our unit tests.                                                                |
+| Boost C++ Thread Library        | 1.33.0 (Mac OS-X only)        | Boost.org thread library is required on Mac OS-X.                                                                                                                                                                                                                                      |
+| Boost C++ Serialization Library | 1.37.0                        | Boost.org serialization library is required for the type system and the MQueue transport.                                                                                                                                                                                              |
+| GNU gcc / g++ Compilers         | 3.4.0 (Linux/Cygwin/Mac OS X) | gcc.gnu.org Orocos builds with the GCC 4.x series as well.                                                                                                                                                                                                                             |
```
To be verified...

meyerj commented on INSTALL.md:
```diff
@@ -0,0 +1,166 @@
+| Boost C++ Thread Library        | 1.33.0 (Mac OS-X only)        | Boost.org thread library is required on Mac OS-X.                                                                                                                                                                                                                                      |
+| Boost C++ Serialization Library | 1.37.0                        | Boost.org serialization library is required for the type system and the MQueue transport.                                                                                                                                                                                              |
+| GNU gcc / g++ Compilers         | 3.4.0 (Linux/Cygwin/Mac OS X) | gcc.gnu.org Orocos builds with the GCC 4.x series as well.                                                                                                                                                                                                                             |
+| MSVS Compilers                  | 2005                          | One can download the MS VisualStudio 2008 Express edition for free.                                                                                                                                                                                                                    |
```
To be verified...

meyerj commented:
### TODO

- [x] split env-hooks and setup scripts in a separate PR (for 2.9) (https://github.com/orocos-toolchain/orocos_toolchain/pull/33)
- [x] split installation script in a separate PR or replace it by cmake with [ExternalProject](https://cmake.org/cmake/help/latest/module/ExternalProject.html) (https://github.com/orocos-toolchain/orocos_toolchain/pull/34)
- Discard Travis script. The macOS builds have now been added to RTT (https://github.com/orocos-toolchain/rtt/pull/195) and we will add a Travis job to https://github.com/orocos-toolchain/ocl, too, which builds against binaries for the respective RTT version if available. No need to run Travis on this repo separately.
- Discard updated installation manual. This will be in a separate documentation repo or we keep it in https://github.com/orocos-toolchain/rtt.
